rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+> **Quote from the trial of CD**: "We'll have to up our game to check things before deploying them. There are always corner cases you don't expect. We're perfectly capable of checking all our changes."
+
+### Automatic deployments
+
+We will enable automatic deployments for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145), but only when they meet the safety criteria of this RFC. To help with this, a rough audit of "compatibility" can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+- We can immediately enable automatic deployments for apps marked with a ✅.
+- We will undertake work to address all the issues (⚠) for apps marked with a ❌.
+
+> **We will disable CD for Publishing API**. This was only enabled on a trial basis, and requires non-trivial work to bring it up to the new standard. We should avoid setting a false precedent by leaving CD enabled.
+
+The following steps must be taken as part of enabling automatic deployments:
+
+- Restrict merge access to "GOV.UK Production" [[example](https://github.com/alphagov/govuk-saas-config/pull/55)].
+
+- Tag [relevant Smoke tests](https://github.com/alphagov/smokey/blob/5dc36fc78c12838ec1c8070d6fbc535ae5433b89/docs/writing-tests.md#filtering-by-app) to run for the app [[example](https://github.com/alphagov/smokey/pull/708/commits/91a5ce94b7b96d09f5db87e63fbc3e400788de00)].
+
+- Add a note in the [Release](https://github.com/alphagov/release) app to warn about the change.
+
+- Add a PR template to warn about the change [[example](https://github.com/alphagov/smart-answers/blob/3509ba3e68c62a206874c96111628dd65adae689/.github/pull_request_template.md)].
+
+  - Backfill the PR template as a comment on existing PRs.
+
+- Enable automatic deployments in a Puppet PR [[example](https://github.com/alphagov/govuk-puppet/pull/10590)].
+
+  - Comment on how the app meets the criteria of this RFC.
+
+> **We can be flexible with coverage**. We can defer adding internal repo tests if the code has no impact on the running app e.g. rake tasks [[1](https://github.com/alphagov/email-alert-api/blob/59fc71a58317ef2998f2c0ef102020da3ca9df96/lib/tasks/support.rake)]. But we should not defer removing any unused code.
+
+### Slack notifications
+
+We need to ensure we are notified about a failure in the automation, so that the automation is no worse than the current, manual process. This has already been done for the `Deploy_App_Downstream` job [[1](https://github.com/alphagov/govuk-puppet/pull/10532)] [[2](https://github.com/alphagov/govuk-puppet/pull/10601)].
+
+So that we cover the entire deployment pipeline, we will also update [the `Deploy_App` job](https://github.com/alphagov/govuk-puppet/blob/f57335f6027487b5b2d2205a35ceb411776f09ce/modules/govuk_jenkins/templates/jobs/deploy_app.yaml.erb) to post a Slack message in the #govuk-developers Slack channel, whenever it fails due to an unexpected error.
```

Comment:
The main reason for #govuk-developers was for consistency with the existing notification we have in place for "Deploy_App_Downstream". I'm not clear myself on why we need [the existing notification](https://github.com/alphagov/govuk-puppet/pull/10532):

- Either we don't really care when the deployment goes out (we have the badger for that); or
- Or we do care and we will probably want to monitor it in the release app anyway.

It would require some effort to make the existing notifications more personalised, since we'd have to manually look up which PR the deployment came from, and do our own Slack integration to notify the author.

I'm growing to dislike my suggestion, because it will likely give a load of false alarms since it's common for people to manually invoke the "Deploy_App" job, and we'll get notified about e.g. typos in parameters.

On balance, I think the notifications now we have in place are sufficient, and we can always reconsider adding more if we find that deploys are silently failing and this is leading to problems as-yet unknown.

How about that?

Reply:
Yup, that sounds good. I hadn't appreciated that "Deploy_App" would notify #govuk-developers too and I agree with removing that.

I'm not convinced #govuk-developers is the right channel for "Deploy_App_Downstream" but since it is already there it can probably be iterated at a later date if it probes problematic.

Reply:
Agreed, and removed 👍